Avoid copy action for root assemblies

sbomer opened the pull request:
Changes `TrimmerRootAssembly` (and the underlying option `-a`) to root all members of the assembly, without setting the action to `copy`. This allows us to perform branch removal and attribute removal for the assembly.

Fixes https://github.com/dotnet/runtime/issues/95429

sbomer committed:
Avoid copy action for root assemblies

sbomer committed:
Ensure attributes get marked too

Preserve existing marking behavior by letting MarkStep do the same
work that it does for copy assemblies, including marking attributes.

sbomer committed:
Fix wasm test failures

sbomer committed:
Fix ApiCompat errors

sbomer committed:
Fix issue, suppress for trimming but not AOT

sbomer committed:
Mark root assemblies visible to reflection

sbomer committed:
Fix unexpected IL2045

With the change to MarkEntireType that calls MarkType instead of
Annotations.Mark, we are now reaching MarkType for the dependency
from the copy assembly to the attribute it contains, producing an
unexpected IL2045. Silence this by checking the DependencyKind.

sbomer committed:
Disable another test

sbomer committed:
Merge remote-tracking branch 'origin/main' into avoidCopyForRootAssembly

sbomer committed:
Preserve DebuggableAttribute for ApplyUpdateTest

sbomer committed:
Set DebuggerSupport for System.Runtime.Tests

sbomer committed:
Fix bug in constant method analysis

BitVector32.CreateMask(int.MinValue) was incorrectly being treated
as returning constant 1. The IL from the beginning of that method:

    IL_0000:  ldarg.0
    IL_0001:  brtrue.s   IL_0005
    IL_0003:  ldc.i4.1
    IL_0004:  ret

The brtrue.s branch was being analyzed as not taken because the
argument was not 1. The branch should be analyzed as taken for
any non-zero argument.

sbomer committed:
Add testcase for root behaviors

And don't set ForceParse for rooted methods.
ForceParse was preventing the XML substitutions from applying.

This change may make it possible for substitutions in copy assemblies
to be honored...

sbomer committed:
Check warning behavior

lambdageek commented:
> The ApplyUpdateTests are running for mono wasm in Release mode, where DebuggerSupport was set to false by default. This illink change results in DebuggableAttribute being removed, causing test failures, so I set DebuggerSupport to true for these tests to preserve the attribute. Now I am getting failures which I don't know how to diagnose, and I'm not sure whether they're expected. For example, this assert is hitting:
> 
> https://github.com/dotnet/runtime/blob/89af7af9d9f477e7969297aacce10c17c1f29ca0/src/libraries/System.Runtime.Loader/tests/ApplyUpdateTest.cs#L226
> 
> @lambdageek Could you please advise? Would it be better to simply disable these tests in Release mode, or are these failures worth investigating?
> 
> edit: It seems those failures aren't reproing in CI - maybe I was doing something incorrectly locally.

@sbomer The EnC tests depend at runtime on the assemblies built from the projects in the `ApplyUpdate` subdirectory.  Those assemblies need to be compiled by Roslyn with optimizations off and with emitted debug information see https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime.Loader/tests/ApplyUpdate/Directory.Build.props

(specifically when a runtime that supports hot reload tries to apply an update to a particular assembly, that assembly must have `[assembly:System.Diagnostics.DebuggableAttribute(DebuggingModes.DisableOptimizations)]`).

In general the assemblies in the `ApplyUpdate` directory must not be modified by the trimmer at all - otherwise the hot reload deltas will not match the assembly contents and the runtime will do something nonsensical. (in real world usage the deltas are generated by VS; for the tests it's the `*.dll.1.dmeta` and `*.dll.1.dil` files that are generated during the test build ahead of time).  You can think of the hot reload deltas as some kind of log that says "add a new row N to table X with contents ABCD; modify row M of table Y with contents SPQR; etc" - if the trimmer disturbs the baseline assembly at all - things will go wrong.

sbomer committed:
Preserve substitution behavior for 'copy' action

sbomer committed:
Use LogDoesNotContain instead

FIxing the checkRemainingErrors behavior results in
too many spurious failures. Use
LogDoesNotContainAttribute for now.

sbomer committed:
Ignore testcase for nativeaot

sbomer committed:
Preserve EnC assemblies for System.Runtime.Loader.Tests

sbomer committed:
Add KeptAttributeAttribute

sbomer committed:
Merge remote-tracking branch 'origin/main' into avoidCopyForRootAssembly

MichalStrehovsky commented on src/libraries/System.Runtime.Loader/tests/System.Runtime.Loader.Tests.csproj:
```diff
@@ -2,6 +2,7 @@
   <PropertyGroup>
     <RootNamespace>System.Runtime.Loader.Tests</RootNamespace>
     <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
+    <DebuggerSupport Condition="'$(DebuggerSupport)' == '' and '$(TargetOS)' == 'browser'">true</DebuggerSupport>
```
Is this still needed with the target below?

sbomer committed:
Remove unnecessary DebuggerSupport setting

MichalStrehovsky commented on src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Attributes.cs:
```diff
@@ -327,6 +328,8 @@ public static class GetCustomAttribute
     {
 
         [Fact]
+        [ActiveIssue("https://github.com/dotnet/linker/issues/2078", typeof(PlatformDetection), nameof(PlatformDetection.IsTrimmedWithILLink))
```
Wondered why this wasn't hit with naot and found out illinker has this as a hardcoded list. Weird, but okay I guess. Would be nice to move this to the same plan as all the other attributes at least. I'm not convinced anyone will want to work on 2078 anytime soon.

MichalStrehovsky commented on src/tools/illink/src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs:
```diff
@@ -1583,7 +1583,7 @@ public bool Analyze (in CalleePayload callee, Stack<MethodDefinition> callStack)
 
 							if (operand is int oint) {
-								if (oint == 1)
+								if (oint != 0)
```
Why was this needed? It's a good change, but surprised to see this in this PR.

MichalStrehovsky reviewed (approved):
Looks reasonable to me although I'm not an expert on this codebase.
